### PR TITLE
INS-23 relay autodiscovery

### PR DESCRIPTION
**Pull request by ooki2day**
**- What I did**

Implemented mechanic for autodetecting relay node. 

**- How I did it**

A node at the start collects some information about another known nodes and uses it for detecting subnets and relaying possibilities

**- How to verify it**

See dht_test.go (func TestDHT_RelayRequest)

### CODE REVIEW

File: network/host/dht_test.go [hunk 1]
```diff
@@ -1095,3 +1104,249 @@ func TestNewDHT(t *testing.T) {
...
+		dhts = append(dhts, dht)
+		assert.NoError(t, err)
+	}
+
+	for _, dht := range dhts {
+		ctx, _ := NewContextBuilder(dht).SetDefaultNode().Build()
+		assert.Equal(t, 0, dht.NumNodes(ctx))
+		go func(dht *DHT) {
+			err := dht.Listen()
+			assert.Equal(t, "closed", err.Error())
```
Comment (AndreyBronin) [inline network/host/dht_test.go]: use const **closedMessage** coz kcpTransport use "broken pipe" message for connection close

File: network/host/dht_test.go [hunk 2]
```diff
@@ -1095,3 +1104,249 @@ func TestNewDHT(t *testing.T) {
...
+		}(dht)
+	}
+
+	for _, dht := range dhts {
+		dht.Bootstrap()
+	}
+
+	ctx, _ := NewContextBuilder(dhts[0]).SetDefaultNode().Build()
+
+	err := dhts[0].AuthenticationRequest(ctx, "wrong command", ids[1])
```
Comment (AndreyBronin) [inline network/host/dht_test.go]: You can use table driven tests for this case

File: network/host/dht_test.go (hunk 1 above)
Comment (ooki2day) [reply network/host/dht_test.go]: fixed

File: network/host/dht_test.go (hunk 2 above)
Comment (ooki2day) [reply network/host/dht_test.go]: fixed

File: network/host/relay/relay_test.go [hunk 3]
```diff
@@ -25,6 +25,31 @@ import (
...
+func TestNewProxy(t *testing.T) {
+	proxy := NewProxy()
+
+	check := true
+
+	if proxy == nil {
+		check = false
+	}
+
+	assert.Equal(t, true, check)
```
Comment (AndreyBronin) [inline network/host/relay/relay_test.go]: assert.NotNil(proxy)

Comment (ooki2day) [reply network/host/relay/relay_test.go]: fixed

**Pull request merged**